### ISSUE TITLE
transpileModule() doesn't catch duplicate let/const

### ISSUE THREAD

**Issue by fatcerberus** [Microsoft/TypeScript#14542]
**TypeScript Version:**  `typescript@next` (2.3.0-dev.20170308)

TypeScript does not seem to produce an error diagnostic for duplicate declarations **when using `transpileModule()`**.  This causes my tooling for [miniSphere](https://github.com/fatcerberus/minisphere) (which uses TS for ES6->ES5 transpilation) to produce possibly incorrect JS in the presence of a duplicate declaration rather than an early error, as happens with Babel.

`tsc` correctly produces an error, so this seems to be a bug in `transpileModule()`:
```
C:\temp>tsc pig.ts
pig.ts(1,5): error TS2451: Cannot redeclare block-scoped variable 'a'.
pig.ts(2,7): error TS2451: Cannot redeclare block-scoped variable 'a'.
```

**Code**

```ts
let a = 1;
const a = 2;
if (true) {
    let a = 3;
}
```

Compiled using:
```js
var output = ts.transpileModule(input, {
	fileName: inFileNames[0],
	reportDiagnostics: true,
	compilerOptions: {
		target: ts.ScriptTarget.ES5,
		module: ts.ModuleKind.CommonJS,
		allowJs: true,
		alwaysStrict: true,
		downlevelIteration: true,
		newLine: ts.NewLineKind.LineFeed,
		noImplicitUseStrict: false,
	},
});
```

**Expected behavior:**
Error diagnostic produced due to duplicate declaration.

**Actual behavior:**
No error diagnostic, produces the following JS:

```js
var a = 1;
var a = 2;
if (true) {
    var a_1 = 3;
}
```

**Comment by RyanCavanaugh** [comment]
`transpileModule`, as the name implies, is only for transpilation. It does not produce semantic diagnostics.

 See https://github.com/Microsoft/TypeScript/wiki/Using-the-Compiler-API

> Creating a compiler is simple enough, but you may want to **just** get the corresponding JavaScript output given TypeScript sources. For this you can use ts.transpileModule 

That page also demonstrates how to set up a full checker pipeline to get semantic diagnostics.

**Comment by fatcerberus** [comment]
I assumed duplicate declarations would be caught, since that's a static syntax error (I think?).  I know I don't get type checking, etc.

**Comment by RyanCavanaugh** [comment]
It's not a *syntax* error - the program conforms to the grammar. It is an *early* error but those are not guaranteed to be found without typechecking.

In general you shouldn't use `transpileModule` if you care about errors of any sort. It's purely for text-to-text transformation.

**Comment by fatcerberus** [comment]
The main thing stopping me from using a full compiler harness is that TS wants to see the complete program including all source files on each build, which prevents me from doing incremental builds.

To clarify, what I need is a file-level ES6->ES5 transformation (TS is *much* faster at this than Babel) without the compile-time module resolution, typechecking, etc. but preferably _with_ early errors.  Is it possible to do that with a full Program harness?

**Comment by RyanCavanaugh** [comment]
That's not possible due to how the checker is designed.

**Comment by fatcerberus** [comment]
Thanks, that means I'll need to rethink my TypeScript integration strategy going forward.  For now I'll get by with `transpileModule()`.

Thank you for the help! 👍

**Comment by fatcerberus** [comment]
> I'm assuming you meant "without a full Program harness" - obviously possible *with*.

Oh, didn't see this edit.  That wasn't a typo, I indeed meant "with" (and it wasn't obvious to me due to the Compiler API documentation being... somewhat lacking).  Good to know!

**Comment by RyanCavanaugh** [comment]
FWIW if your build process is resident, it's very cheap once everything is set up to update a new Program instance, typecheck a specific file, and get its emit. This is how the language service / compile-on-save works in editors.

**Comment by fatcerberus** [comment]
Yeah, unfortunately my build process is just an SCons-like command-line tool, so watching for changes in realtime isn't an option for me right now. 🙁

**Comment by fatcerberus** [comment]
Thanks for the help Ryan, I can close this now since there's nothing to fix.
